File: core/src/commonTest/kotlin/io/silencelen/andvari/core/client/StrengthTest.kt
package io.silencelen.andvari.core.client

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertNotNull
import kotlin.test.assertNull
import kotlin.test.assertTrue

/** Pinned scores mirroring web/src/ui/strength.ts — the two impls must agree exactly,
 *  or the spec 07 §2.3 backup-passphrase floor (score ≥ 3) drifts between clients. */
class StrengthTest {
    @Test
    fun pinnedScores_matchWebEstimator() {
        assertEquals(0, Strength.estimateStrength(""))                        // 0 bits
        assertEquals(0, Strength.estimateStrength("abcdefgh"))                // an 8-long ±1 run → 2 effective × 2
        assertEquals(0, Strength.estimateStrength("aaaaaaaaaaaaaaaaaaaa"))    // one run → 2 effective × 2 (F31: was 20×2 = 40)
        assertEquals(2, Strength.estimateStrength("Tr0ub4dor&3"))             // 11×6 = 66
        assertEquals(2, Strength.estimateStrength("correct horse battery"))   // 21×3.5 = 73.5 (2 classes: lower+space)
        assertEquals(1, Strength.estimateStrength("Aa1!Aa1!Aa"))              // "Aa1!"×2 + "Aa" → 7×6 = 42 (F31: was 60)
        assertEquals(1, Strength.estimateStrength("Aa1!Aa1!Aa1!Aa1"))         // "Aa1!"×3 + "Aa1" → 8×6 = 48 (F31: was 90)
        assertEquals(2, Strength.estimateStrength("Abcdefgh12Abcdefgh12Ab"))  // block×2 + "Ab" → 13×5 = 65 (F31: was 110)
        assertEquals(0, Strength.estimateStrength("Aa1!Aa1!Aa1!Aa1!Aa1!"))    // 5×"Aa1!" → 5×6 = 30 (F31: was 120)
        // Unpatterned passphrases are untouched by F31 — the penalty must not tax real ones.
        assertEquals(3, Strength.estimateStrength("correct horse battery staple"))
        assertEquals(4, Strength.estimateStrength("the quick brown fox jumps over the lazy dog 42 TIMES!"))
    }

    /**
     * F31: the whole point — a doubled password no longer scores as double the entropy.
     * `Password1!Password1!` was the audit's example of "strong" (4) for a 10-character choice.
     */
    @Test
    fun repetitionAndSequences_collapseBeforeTheLengthTerm() {
        assertEquals(2, Strength.estimateStrength("Password1!Password1!"))    // the token + 1, not the span
        assertEquals(4, Strength.entropyProxyScore("Password1!Password1!"))   // what it used to score
        assertEquals(0, Strength.estimateStrength("a".repeat(40)))            // 40-char run: audit's other example
        assertEquals(3, Strength.entropyProxyScore("a".repeat(40)))
        assertEquals(0, Strength.estimateStrength("ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklm")) // two ±1 runs
        // The smallest period wins: "abab…" collapses as "ab", never as "abab".
        assertEquals(Strength.estimateStrength("ababababababababab"), Strength.estimateStrength("abababababab"))
        // Sub-threshold coincidences a random generator produces routinely are NOT patterns.
        assertEquals(Strength.entropyProxyScore("Xk7#mQ2\$vL9!pR4&"), Strength.estimateStrength("Xk7#mQ2\$vL9!pR4&"))
        assertEquals(Strength.entropyProxyScore("correct-horse-battery-staple"), Strength.estimateStrength("correct-horse-battery-staple"))
    }

    /**
     * F31 owner decision: the pattern penalty WARNS, it never refuses. Every password that
     * cleared the master-password floor before F31 still clears it — otherwise a household
     * member could be locked out of *changing* their password by a stricter estimator.
     */
    @Test
    fun masterPasswordFloor_neverTightenedByThePatternPenalty() {
        for (pw in listOf(
            "Password1!Password1!",
            "a".repeat(40),
            "aA1!aA1!aA1!aA1!aA1!",
            "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklm",
            "correct-horse-battery-staple",
        )) {
            assertTrue(Strength.meetsMasterPasswordFloor(pw), "F31 must not refuse '$pw'")
            assertEquals(Strength.entropyProxyScore(pw) >= Strength.MASTER_PW_MIN_SCORE, Strength.meetsMasterPasswordFloor(pw))
        }
        // …and it still refuses what F60 always refused.
        assertFalse(Strength.meetsMasterPasswordFloor("password"))
        assertFalse(Strength.meetsMasterPasswordFloor(""))
    }

    @Test
    fun patternWarning_firesOnlyWhenThePatternIsWhatMakesItWeak() {
        assertNotNull(Strength.patternWarning("Password1!Password1!"))
        assertEquals(Strength.PATTERN_WARNING, Strength.patternWarning("a".repeat(40)))
        assertTrue(Strength.hasPatternWeakness("a".repeat(40)))
        assertNull(Strength.patternWarning("correct-horse-battery-staple"))
        assertNull(Strength.patternWarning(""))
        assertFalse(Strength.hasPatternWeakness(""))
    }

    @Test
    fun labels_matchWeb() {
        assertEquals(listOf("very weak", "weak", "fair", "good", "strong"), Strength.LABELS)
        assertEquals("very weak", Strength.label(0))
        assertEquals("good", Strength.label(Strength.BACKUP_FLOOR))
        assertEquals("strong", Strength.label(4))
        assertEquals("strong", Strength.label(9)) // clamped
    }
}

File: web/src/ui/strength.test.ts
import { describe, expect, it } from "vitest";
import {
  MASTER_PW_MIN_SCORE,
  PATTERN_WARNING,
  entropyProxyScore,
  estimateStrength,
  hasPatternWeakness,
  masterPasswordHasNonAscii,
  meetsMasterPasswordFloor,
  patternWarning,
} from "./strength";

/**
 * F60 (spec 05 T8 / spec 01 §1): the master-password floor that gates enrollment + every
 * password change (Welcome + Settings). Pinned so a future estimateStrength tweak can't
 * silently drop the vault-wrapping password below what a backup export already demands.
 */
describe("master-password floor (F60)", () => {
  it("the floor is score ≥ 3 (\"good\")", () => {
    expect(MASTER_PW_MIN_SCORE).toBe(3);
  });

  it("rejects the passwords the OLD length≥8 gate let through", () => {
    // 8+ chars but a single character class → weak; the whole point of F60.
    expect(meetsMasterPasswordFloor("password")).toBe(false);
    expect(meetsMasterPasswordFloor("aaaaaaaaaa")).toBe(false);
    expect(meetsMasterPasswordFloor("12345678")).toBe(false);
    // Short-but-mixed is still below the bit floor.
    expect(meetsMasterPasswordFloor("Ab3$")).toBe(false);
  });

  it("accepts genuinely strong passwords", () => {
    expect(meetsMasterPasswordFloor("correct-horse-battery-staple")).toBe(true);
    expect(meetsMasterPasswordFloor("Tr0ub4dor&3xtra-Length!!")).toBe(true);
    expect(meetsMasterPasswordFloor("a-fairly-long-diceware-style-phrase")).toBe(true);
  });

  it("gates on entropyProxyScore — F31's pattern penalty warns, it never refuses", () => {
    for (const pw of ["password", "Ab3$", "correct-horse-battery-staple", ""]) {
      expect(meetsMasterPasswordFloor(pw)).toBe(entropyProxyScore(pw) >= 3);
    }
    // The audit's two examples: both now SCORE lower, and both still clear the floor, because
    // no household member may be locked out of changing their password by a stricter estimator.
    for (const pw of ["Password1!Password1!", "a".repeat(40), "aA1!aA1!aA1!aA1!aA1!"]) {
      expect(estimateStrength(pw), pw).toBeLessThan(entropyProxyScore(pw));
      expect(meetsMasterPasswordFloor(pw), pw).toBe(true);
    }
  });

  it("empty never clears the floor (submit guards must still block it)", () => {
    expect(meetsMasterPasswordFloor("")).toBe(false);
  });

  it("flags non-ASCII (a SHOULD-warn, not a block)", () => {
    expect(masterPasswordHasNonAscii("plain-ascii-123")).toBe(false);
    expect(masterPasswordHasNonAscii("café-passphrase")).toBe(true); // é
    expect(masterPasswordHasNonAscii("naïve-☺-secret")).toBe(true);
    // The warn is independent of the floor — a strong non-ASCII password still passes.
    expect(meetsMasterPasswordFloor("café-très-long-passphrase-2024")).toBe(true);
  });
});

/**
 * F31: `length × classes` scored a doubled password as twice the entropy —
 * `Password1!Password1!` rendered as "strong" and a 40-character run of one letter cleared
 * both floors. Repeated blocks and ±1 runs now collapse before the length term. Kept in
 * lockstep with core Strength.kt (StrengthTest) and spec/test-vectors/strength.json.
 */
describe("pattern collapse (F31)", () => {
  it("charges a repeated block as the token plus one, not the whole span", () => {
    expect(estimateStrength("Password1!Password1!")).toBe(2);
    expect(entropyProxyScore("Password1!Password1!")).toBe(4); // what it used to score
    expect(estimateStrength("aA1!aA1!aA1!aA1!aA1!")).toBe(0);
    expect(estimateStrength("a".repeat(40))).toBe(0);
    // The smallest period wins: "abab…" collapses as "ab", never as "abab".
    expect(estimateStrength("ababababababababab")).toBe(estimateStrength("abababababab"));
  });

  it("charges an ascending/descending run as 2", () => {
    expect(estimateStrength("ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklm")).toBe(0);
    expect(estimateStrength("abcdefgh")).toBe(0);
  });

  it("leaves real passphrases and generated passwords alone", () => {
    for (const pw of [
      "correct horse battery staple",
      "correct-horse-battery-staple",
      "the quick brown fox jumps over the lazy dog 42 TIMES!",
      "Xk7#mQ2$vL9!pR4&",
      "Tr0ub4dor&3xtra-Length!!",
    ]) {
      expect(estimateStrength(pw), pw).toBe(entropyProxyScore(pw));
      expect(patternWarning(pw), pw).toBeNull();
    }
  });

  it("warns only when the pattern is what makes it weak", () => {
    expect(patternWarning("Password1!Password1!")).toBe(PATTERN_WARNING);
    expect(hasPatternWeakness("a".repeat(40))).toBe(true);
    expect(patternWarning("")).toBeNull();
    expect(hasPatternWeakness("")).toBe(false);
  });
});
